common/core/desktop/src/kmx/kmx_debugger.cpp

```diff
@@ -11,12 +11,56 @@
 using namespace km::kbp;
 using namespace kmx;
 
-void KMX_DebugItems::push_item(uint8_t type, uint32_t flags, LPGROUP group, LPKEY key, PKMX_WCHAR context) {
+void KMX_DebugItems::push_item(uint8_t type, uint32_t flags, LPGROUP group, LPKEY key, PKMX_WCHAR context, PKMX_WORD index_stack) {
   _items->assert_push_entry();
   km_kbp_state_debug_item item = {type, flags};
   item.kmx_info.rule = key;
-  u16ncpy(item.kmx_info.context, context, DEBUG_MAX_CONTEXT - 1);
+  if(item.kmx_info.rule && index_stack) {
+    this->fill_store_offsets(&item.kmx_info, index_stack);
+  }
+
+  if(context != nullptr) {
+    u16ncpy(item.kmx_info.context, context, DEBUG_MAX_CONTEXT - 1);
+  }
   item.kmx_info.context[DEBUG_MAX_CONTEXT-1] = 0;
   item.kmx_info.group = group;
   _items->emplace_back(item);
 }
+
+void KMX_DebugItems::fill_store_offsets(km_kbp_state_debug_kmx_info *info, PKMX_WORD index_stack) {
```

Comment:
Yeah... parsing this particular function may be beyond me, at least without additional context.

'course, I'm not the usual reviewer for these, so maybe someone else can handle it on my behalf.

Reply:
I hope to improve this function; right now it's essentially a clone of the keyman32 code and I didn't want to clean it up until I had the basic framework going within both Core and Keyman Developer.

Reply:
Yes I can see this replicating FillStoreOffsets from `aiDebugger.cpp`  however I don't follow what IndexStack is doing here. So I can say yes the behaviour is copied.

Reply:
> however I don't follow what IndexStack is doing here

`IndexStack` is keeping track of the matched character index within a store, e.g.:

```
store(foo) 'abc'
store(bar) 'ABC'

any(foo) + 'x' > index(bar, 1)
```

If you then type 'cx', `IndexStack` would have the value `2` for `any(foo)` (as 'c' is third character in the store, 0-based index).